### PR TITLE
add runtime_conditional and flow unary operator*

### PR DESCRIPTION
**Pull request by lukevalenty**
TODO:
- ✅Better way to ensure predicates for the same condition are unique
- ✅operator&& for predicates
- ✅Better error message for edge/node predicate requirements
- ✅Documentation
- ✅More unit tests for edge/node predicate requirements
- ✅unary operator* for seq/par objects 
- ~~unary operator- for explicit references (can't be made into non-reference by unary +)~~
- ✅Message handler support
- ✅Indexed message handler tests
- ~~runtime_condition support for cib::callback~~ (do in follow-on PR)

### CODE REVIEW

**Comment by lukevalenty** [comment]
Passes CI in our downstream repo. 😎

File: include/flow/graph_builder.hpp [hunk 1]
```diff
@@ -119,13 +187,79 @@ struct graph_builder {
...
+        if constexpr (not std::is_same_v<node_names_t,
+                                         mentioned_node_names_t>) {
+            using missing_nodes_t =
+                boost::mp11::mp_set_difference<mentioned_node_names_t,
+                                               node_names_t>;
+
+            // convert missing_nodes_t into an mp list
+            using missing_nodes_tuple_t =
+                boost::mp11::mp_apply<stdx::tuple, missing_nodes_t>;
+            constexpr auto missing_nodes_tuple = missing_nodes_tuple_t{};
```
Comment (bdeane-intel) [inline include/flow/graph_builder.hpp:210]: **issue**: `missing_nodes_t` is already a tuple? (and the comment is wrong)

File: include/flow/detail/walk.hpp [hunk 2]
```diff
@@ -47,7 +47,11 @@ constexpr inline class get_finals_t {
 
 constexpr inline class get_nodes_t {
     template <node N> friend constexpr auto tag_invoke(get_nodes_t, N &&n) {
-        return stdx::make_tuple(std::forward<N>(n));
+        if constexpr (std::remove_cvref_t<N>::is_reference) {
```
Comment (bdeane-intel) [inline include/flow/detail/walk.hpp:50]: **thought**: I just noticed how odd this looks...

File: include/flow/graph_builder.hpp [hunk 3]
```diff
@@ -177,10 +311,17 @@ template <stdx::ct_string Name = "", typename Renderer = graph_builder<impl>,
...
         });
-        return stdx::to_unsorted_set(t);
+    }
+
+    friend constexpr auto tag_invoke(flow::dsl::get_all_mentioned_nodes_t,
+                                     graph const &g) {
+        return g.fragments.apply([](auto const &...frags) {
+            return stdx::tuple_cat(
+                flow::dsl::get_all_mentioned_nodes(frags)...);
+        });
```
Comment (bdeane-intel) [inline include/flow/graph_builder.hpp:324]: **suggestion**: Move the `stdx::to_unsorted_set` from the `get_edges_t` overload of `tag_invoke` to `build`, then collapse all three overloads here to:
```cpp
friend constexpr auto tag_invoke(auto tag, graph const &g) {
  return g.fragments.apply([](auto const &...frags) {
      return stdx::tuple_cat(tag(frags)...);
  });
}
```

File: include/flow/detail/walk.hpp (hunk 2 above)
Comment (lukevalenty) [reply include/flow/detail/walk.hpp:50]: anything we should do about it?

Comment (bdeane-intel) [reply include/flow/detail/walk.hpp:50]: rename `is_reference` to `is_pointer`? :)

Comment (bdeane-intel) [reply include/flow/detail/walk.hpp:50]: And find a name that doesn't quite clash with a `std` type_trait...

Comment (lukevalenty) [reply include/flow/detail/walk.hpp:50]: Ohhh....yes, i see it now!

File: include/cib/detail/runtime_conditional.hpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    CONSTEVAL explicit runtime_conditional(Configs const &...configs)
+        : body{configs...} {}
+
+    [[nodiscard]] constexpr auto extends_tuple() const {
+        return stdx::transform(
+            []<typename E>(E e) {
+                auto args_tuple = stdx::transform(
+                    []<typename Arg>(Arg) {
+                        return make_runtime_conditional(Cond{}, Arg{});
```
Comment (elbeno) [inline include/cib/detail/runtime_conditional.hpp:25]: **todo**: Provide a "poison pill" function template in this namespace that is found so that ADL can occur in all cases. (Consider making `make_runtime_conditional` a function object?)

File: include/cib/detail/runtime_conditional.hpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    [[nodiscard]] CONSTEVAL auto operator()(Configs const &...configs) const {
+        return detail::runtime_conditional<runtime_condition<Name, Ps...>,
+                                           Configs...>{configs...};
+    }
+
+    explicit operator bool() const { return (Ps{}() and ...); }
+
+    template <stdx::ct_string RhsName, typename... RhsPs>
+    [[nodiscard]] constexpr auto
+    operator and(runtime_condition<RhsName, RhsPs...> rhs) const {
```
Comment (elbeno) [inline include/cib/detail/runtime_conditional.hpp:60]: **suggestion**: Make this a free binary function

File: include/flow/detail/walk.hpp [hunk 6]
```diff
@@ -8,15 +10,16 @@
 
 namespace flow::dsl {
 template <typename T>
-concept node = requires { typename stdx::remove_cvref_t<T>::is_node; };
+concept subgraph = requires { typename stdx::remove_cvref_t<T>::is_subgraph; };
```
Comment (elbeno) [inline include/flow/detail/walk.hpp:13]: **suggestion**: Follow-up with beefing up this concept, and removing `tag_invoke`. I like the name change from `node` to `subgraph`.

File: include/flow/detail/walk.hpp [hunk 7]
```diff
@@ -8,15 +10,16 @@
...
 
-template <typename Source, typename Dest> struct edge {
+template <typename Source, typename Dest, typename Cond> struct edge {
```
Comment (elbeno) [inline include/flow/detail/walk.hpp:15]: **suggestion**: Expose a `cond_t` as well as the `source_t` and `dest_t`. (I know the condition is used by deducing/destructuring the type, but for consistency's sake...)

File: include/flow/detail/walk.hpp [hunk 8]
```diff
@@ -46,8 +50,13 @@ constexpr inline class get_finals_t {
 } get_finals{};
 
 constexpr inline class get_nodes_t {
-    template <node N> friend constexpr auto tag_invoke(get_nodes_t, N &&n) {
-        return stdx::make_tuple(std::forward<N>(n));
+    template <subgraph N> friend constexpr auto tag_invoke(get_nodes_t, N &&n) {
+        if constexpr (std::remove_cvref_t<N>::identity ==
+                      subgraph_identity::REFERENCE) {
```
[the remark this answers is not in the document]
Comment (lukevalenty) [reply include/flow/detail/walk.hpp:55]: It started out like that, but some earlier versions of clang didn't like me using `n` in a constant expression :(

File: include/flow/detail/walk.hpp (hunk 6 above)
Comment (lukevalenty) [reply include/flow/detail/walk.hpp:13]: we'll do this in a follow-on PR

**Review by mjcaisse-intel — CHANGES_REQUESTED** [review]

File: docs/flows.adoc [hunk 9]
```diff
@@ -58,6 +58,17 @@ a shower. The flow library will order actions in a flow to respect these
 dependencies. The actions will be executed in an order that respects all given
 dependencies.
 
+The `*` operator is used to explicitly add an action to the
```
[the remark this answers is not in the document]
Comment (lukevalenty) [reply docs/flows.adoc:61]: I need to write the documentation for core cib in order to do this since it's not specific to flows. I can do it as part of this PR, or another follow-on PR. It will take me some time.

[the remark this answers is not in the document]
Comment (bdeane-intel) [reply docs/flows.adoc:61]: I move to do that as a follow-up

**Review by mjcaisse-intel — APPROVED** [review]
